apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js

```diff
@@ -0,0 +1,257 @@
+ *
+ * @typedef { import('./').Attributes } Attributes
+ * @typedef { Object } Props
+ * @property { Color } backgroundButtonColor
+ * @property { Color } textButtonColor
+ * @property { string } fallbackBackgroundColor
+ * @property { string } fallbackTextColor
+ * @property { Attributes } attributes
+ * @property { () => void } setBackgroundButtonColor
+ * @property { () => void } setTextButtonColor
+ * @property { (attributes: Partial<Attributes>) => void } setAttributes
+ *
+ * @param { Props } props
+ */
+function Edit( props ) {
+    useEffect( () => {
+        setButtonClasses();
+        setCustomButtonColors();
+        setBlockContext();
+    }, [ props.backgroundButtonColor, props.textButtonColor, props.context ] );
+
+    function setBlockContext() {
+        const { context } = props;
+        props.setAttributes( context );
+    }
+
+    function setButtonClasses() {
+        const buttonClasses = getButtonClasses();
+        props.setAttributes( { buttonClasses } );
+    }
+
+    function setCustomButtonColors() {
+        const customTextButtonColor = getTextButtonColor();
+        const customBackgroundButtonColor = getBackgroundButtonColor();
+
+        if ( customTextButtonColor !== undefined ) {
+            props.setAttributes( { customTextButtonColor } );
+        }
+
+        if ( customBackgroundButtonColor !== undefined ) {
+            props.setAttributes( { customBackgroundButtonColor } );
+        }
+    }
+
+    function getTextButtonColor() {
+        return get( props.textButtonColor, 'color' );
+    }
+
+    function getBackgroundButtonColor() {
+        return get( props.backgroundButtonColor, 'color' );
+    }
+
+    function getButtonClasses() {
+        const { textButtonColor, backgroundButtonColor } = props;
+        const textClass = get( textButtonColor, 'class' );
+        const backgroundClass = get( backgroundButtonColor, 'class' );
+        return classnames( 'wp-block-button__link', {
+            'has-text-color': textButtonColor.color,
+            [ textClass ]: textClass,
+            'has-background': backgroundButtonColor.color,
+            [ backgroundClass ]: backgroundClass,
+        } );
+    }
+
+    const {
+        isSelected,
+        attributes,
+        setAttributes,
+        backgroundButtonColor,
+        textButtonColor,
+        setBackgroundButtonColor,
+        setTextButtonColor,
+        fallbackBackgroundColor,
+        fallbackTextColor,
+    } = props;
+
+    const backgroundColor = backgroundButtonColor.color || fallbackBackgroundColor;
+    const color = textButtonColor.color || fallbackTextColor;
+    const buttonStyle = { border: 'none', backgroundColor, color };
+    const buttonClasses = getButtonClasses();
+
+    return [
+        isSelected && blockControls( props ),
+
+        isSelected && inspectorControls( props ),
+
+        <div className={ props.className }>
+            <div className="wp-block-button premium-content-logged-out-view-button">
```

Comment:
If we could save this block in the client, it should be possible to re-use the `core/button` blocks instead of reimplementing them here:

```
const TEMPLATE = [
  [ 'core/button', { 
    text: attributes.buttonType === 'login' ? 'Log in' : 'Subscribe',
    url: attributes.buttonType === 'login' ? accessUrl : subscribeUrl, // We need to expose those URLs in the client somehow.
  } ]
];
```

```
<InnerBlocks
  template={ TEMPLATE }
  templateLock="insert"
/>
```

With `templateLock="insert"` we make sure blocks cannot be inserted or removed, but existing blocks provided by the template can be updated.

Reply:
Realized it's better to implement our own blocks rather than re-using `core/block`: https://github.com/Automattic/wp-calypso/pull/42715#discussion_r433197154